### PR TITLE
NAS-128754 / 24.10 / Refactor SMB and Directory Services plugins

### PR DESCRIPTION
**Pull request by anodos325**
These plugins and their various methods still had vestigial code and design from when these plugins were clustered. This commit simplifies general SMB configuration (flat file for global settings rather than using the registry).

This commit moves registry-related functions to dedicated utils that can be imported and directly called.

Methods using a samba loadparm context are moved into a dedicated utils file as well and protected under a lock.

Various constants are moved into a dedicated file as well.

Improved error handling is added for wrappers around "net conf" which uses libsmbconf to effect changes in the Samba running configuration. Specifically, there are edge cases where user can use auxiliary parameters to ram garbage into our configuration. If libsmbconf rejects the auxiliary parameters, we should roll back to a clean state and raise a ValidationError.

An initial round of tests for parser that converts from our TrueNAS configuration to smb.conf parameters is also added.

### CODE REVIEW

**Review by bmeagherix — APPROVED** [review]

**Review by mgrimesix — APPROVED** [review]

File: src/middlewared/middlewared/etc_files/local/smb4.conf.mako [hunk 1]
```diff
@@ -1,17 +1,300 @@
...
+
+    if smb_bindips := render_ctx['smb.config']['bindip']:
+       allowed_ips = set(middleware.call_sync('smb.bindip_choices').values())
+       if (rejected := set(smb_bindips) - allowed_ips):
+           middleware.logger.warning(
+               '%s: IP address(es) are no longer in use and should be removed '
+               'from SMB configuration.', rejected
+           )
+
+       smbconf['interfaces'] = ' '.join(allowed_ips & set(smb_bindips))
```
[the remark this answers is not in the document]
Comment (anodos325) [reply src/middlewared/middlewared/etc_files/local/smb4.conf.mako:150]: Yes. I think next PR I'll add some alerts.

File: src/middlewared/middlewared/etc_files/local/smb4.conf.mako [hunk 2]
```diff
@@ -1,17 +1,300 @@
...
+                case 'ssl':
+                    if v != 'STARTTLS':
+                        disable_startls = True
+                    continue
+                case _:
+                    value = v
+
+            smbconf.update({f'{idmap_prefix} {backend_parameter}': value})
+
+%>\
```
[the remark this answers is not in the document]
Comment (anodos325) [reply src/middlewared/middlewared/etc_files/local/smb4.conf.mako:291]: That's a good idea.

File: src/middlewared/middlewared/plugins/idmap.py [hunk 3]
```diff
@@ -876,7 +877,7 @@ async def do_update(self, id_, data):
         )
 
         out = await self.query([('id', '=', id_)], {'get': True})
-        await self.synchronize(False)
+        await self.middleware.call('etc.generate', 'smb')
```
[the remark this answers is not in the document]
Comment (anodos325) [reply src/middlewared/middlewared/plugins/idmap.py:880]: the `False` passed here skipped the restart.

[the remark this answers is not in the document]
Comment (bmeagherix) [reply src/middlewared/middlewared/plugins/idmap.py:880]: Duh.  Sorry, I got copy/paste trigger happy.

**Pull request merged**